src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,399 @@
+                            break;
+                    }
+
+                    yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+
+                    // Skip the separator.
+                    ichCur++;
+                }
+            }
+            else
+            {
+                for (int ichCur = ichMin; ;)
+                {
+                    int ichMinLocal = ichCur;
+                    for (; ; ichCur++)
+                    {
+                        Contracts.Assert(ichCur <= ichLim);
+                        if (ichCur >= ichLim)
+                        {
+                            yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+                            yield break;
+                        }
+                        // REVIEW: Can this be faster?
+                        if (ContainsChar(memory.Span[ichCur], separators))
+                            break;
+                    }
+
+                    yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+
+                    // Skip the separator.
+                    ichCur++;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Splits this instance on the left-most occurrence of separator and produces the left
+        /// and right ReadOnlyMemory values. If this instance does not contain the separator character,
+        /// this returns false and sets <paramref name="left"/> to this instance and <paramref name="right"/>
+        /// to the default ReadOnlyMemory value.
+        /// </summary>
+        public static bool SplitOne(char separator, out ReadOnlyMemory<char> left, out ReadOnlyMemory<char> right, ReadOnlyMemory<char> memory)
+        {
+            if (memory.IsEmpty)
+            {
+                left = memory;
+                right = default;
+                return false;
+            }
+
+            int ichMin = 0;
+            int ichLim = memory.Length;
+            var text = memory.Span;
+            int ichCur = ichMin;
+            for (; ; ichCur++)
+            {
+                Contracts.Assert(ichMin <= ichCur && ichCur <= ichLim);
+                if (ichCur >= ichLim)
+                {
+                    left = memory;
+                    right = default;
+                    return false;
+                }
+                if (text[ichCur] == separator)
+                    break;
+            }
+
+            // Note that we don't use any fields of "this" here in case one
+            // of the out parameters is the same as "this".
+            left = memory.Slice(ichMin, ichCur - ichMin);
+            right = memory.Slice(ichCur + 1, ichLim - ichCur - 1);
+            return true;
+        }
+
+        /// <summary>
+        /// Splits this instance on the left-most occurrence of an element of separators character array and
+        /// produces the left and right ReadOnlyMemory values. If this instance does not contain any of the
```

Comment:
>ReadOnlyMemory [](start = 40, length = 14)

Another textual reference to a non-generic `ReadOnlyMemory` structure. I think these can be replaced more or less well with `<see cref="ReadOnlyMemory{char}"/>`.

Reply:
<see cref="ReadOnlyMemory{char}"/>  gives an error but apparently it needs to be <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/>

---
In reply to: [216127057](https://github.com/dotnet/machinelearning/pull/863#discussion_r216127057) [](ancestors = 216127057)